docs/poc2023q1/outline.md

```diff
@@ -282,12 +282,169 @@ this category.
+Next, take the union of [the preferred
+versions](https://github.com/kubernetes/apimachinery/blob/release-1.24/pkg/apis/meta/v1/types.go#L1054).
+If this union has a non-empty intersection with the intersection of
+the supported versions, take the following steps with this
+intersection; otherwise proceed with just the intersection of the
+supported versions.  When first (since process startup) presented with
+an instance of this problem, the placement translator picks the
+highest version from this intersection.  Subsequently for the same API
+group, the placement translator sticks with its previous decision as
+long as that is still in the intersection.  If the previous choice is
+no longer avaiable, the highest version is picked.  This preference
+for highest version is based on the expectation that rolling forward
+will be more common than rolling back; using the intersection ensures
+that both work (as long as the collection of sources has an overlap in
+supported versions, which is basic sanity).
+
+A workload prescription object that is in the process of graceful
+deletion (i.e., with `DeletionTimestamp` set to something) is
+considered here to already be gone.
+
+Once they have been read at a consistent API version, merging of
+multiple objects is done as follows.  Different parts of the object
+are handled differently.
+
+- **TypeMeta**.  This can not conflict because it is part of what
+  identifies an object.
+- **ObjectMeta**.
+  - **Labels and Annotations**.  These are handled on a key-by-key
+    basis.  Distinct keys do not conflict.  When multiple objects have
+    a label or annotation with the same key, the corresponding value
+    in the result is the value from the most recently updated of those
+    objects.
+  - **OwnerReferences**.  This is handled analogously to labels and
+    annotations.  The key is the combination of APIVersion, Kind, and
+    Name.
+  - **Finalizers**.  This is simply a set of strings.  The result of
+    merging is the union of the sets.
+  - **ManagedFields**.  This is metadata that is not propagated.
+- **Spec**.  Beyond TypeMeta and ObjectMeta, the remaining object
+  fields are specific to the kind of object.  Many have a field named
+  "Spec" in the Go language source, "spec" in the JSON representation.
+  For objects that have Spec fields, merging has a conflict if those
+  field values are not all equal when considered as JSON data, and the
+  resolution is to take the value from the most recently updated
+  object.
+- **Status**.  Status is handled analogously to Spec.  For both, we
+  consider a missing field to be the same as a field with a value of
+  `nil`.  That is expected to be the common case for the Status of
+  these workload prescription objects.
+- **Other fields**.  If all the values are maps (objects in
+  JavaScript) then they are merged key-by-key, as for labels and
+  annotations.  Otherwise they are treated as monoliths, as for Spec
+  and Status.
+
```

Comment:
A resource does not have a version.  An API group has a _preferred_ version and, in general, _multiple_ versions that it can be read and written as.  So the outline here is to first pick a common version at which to read all the objects that have to be merged together, and after that there is no problem with differing versions.  It would be much harder to define merging of object contents if they are at different versions, so I deliberately did not try to go that way.

Reply:
I think that would be useful. Users may want to know that there was a conflict and how it was managed (and ideally have some input on how it should be managed)

Reply:
You can propose something in a follow-on PR.